src/core/thread/mle_router.cpp

```diff
@@ -184,6 +184,24 @@ Error MleRouter::SetRouterEligible(bool aEligible)
     return error;
 }
 
+// If the router was a leader or had more than 5 children prior to reset,
+// the multicast link request is retransmitted as a critical message
+void MleRouter::SetLinkRequestTransmissionCounter(void)
+{
+    Settings::NetworkInfo networkInfo;
+    uint16_t              numOfChildren = mChildTable.GetNumChildren(Child::kInStateValidOrRestoring);
+    mLinkRequestAttempts                = kMaxTransmissionCount;
+
+    SuccessOrExit(Get<Settings>().Read(networkInfo));
+    if (networkInfo.GetRole() == kRoleLeader || numOfChildren >= kMinCriticalChildrenCount)
```

Comment:
This works fine.

One idea to avoid reading from `Setting` again,  we can track this in `Restore()`
- We already have a boolean `mHasRestored`. 
- We can add another boolean `mWasLeader` and set it in `Restore()` 
- And then use it here to determine the number.